### PR TITLE
Write state.json atomically

### PR DESCRIPTION
**Pull request by mrunalp**
@AkihiroSuda @kolyshkin ptal

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

### CODE REVIEW

File: libcontainer/container_linux.go [hunk 1]
```diff
@@ -1674,13 +1674,36 @@ func (c *linuxContainer) updateState(process parentProcess) (*State, error) {
...
+	defer func() {
+		if retErr != nil {
+			tmpFile.Close()
+			os.Remove(tmpFile.Name())
+		}
+	}()
+
+	err = utils.WriteJSON(tmpFile, s)
+	if err != nil {
+		tmpFile.Close()
```
Comment (kolyshkin) [inline libcontainer/container_linux.go:1692]: No need to close the file on error path as it is done in `defer`

File: libcontainer/container_linux.go [hunk 2]
```diff
@@ -1674,13 +1674,36 @@ func (c *linuxContainer) updateState(process parentProcess) (*State, error) {
...
+		return err
+	}
+	err = tmpFile.Sync()
+	if err != nil {
+		tmpFile.Close()
```
Comment (kolyshkin) [inline libcontainer/container_linux.go:1697]: ditto

**Comment by kolyshkin** [comment]
Is this to prevent reading back a partially written `state.json`?

File: libcontainer/container_linux.go (hunk 1 above)
Comment (mrunalp) [reply libcontainer/container_linux.go]: fixed.

File: libcontainer/container_linux.go (hunk 2 above)
Comment (mrunalp) [reply libcontainer/container_linux.go]: fixed.

**Comment by mrunalp** [comment]
> Is this to prevent reading back a partially written state.json?

Yes! :)

File: libcontainer/container_linux.go [hunk 3]
```diff
@@ -1674,13 +1674,34 @@ func (c *linuxContainer) updateState(process parentProcess) (*State, error) {
...
+			tmpFile.Close()
+			os.Remove(tmpFile.Name())
+		}
+	}()
+
+	err = utils.WriteJSON(tmpFile, s)
+	if err != nil {
+		return err
+	}
+	err = tmpFile.Sync()
```
Comment (AkihiroSuda) [inline libcontainer/container_linux.go:1694]: Do you need this?

Comment (mrunalp) [reply libcontainer/container_linux.go:1694]: Yes, we want to sync the file to make sure it is written to disk.

Comment (mrunalp) [reply libcontainer/container_linux.go:1694]: That applies if the root isn't a tmpfs.

Comment (mrunalp) [reply libcontainer/container_linux.go:1694]: I can drop it if we think we don't want to support that, but I don't see big downsides in keeping it.

**Comment by cyphar** [comment]
Rather than open-coding this, why not ~~steal~~ use [the AtomicFileWriter from Docker et al](https://github.com/moby/moby/blob/master/pkg/ioutils/fswriters.go)?

File: libcontainer/container_linux.go (hunk 3 above)
Comment (AkihiroSuda) [reply libcontainer/container_linux.go:1694]: Sync may block, and this file does not need persistency

Comment (kolyshkin) [reply libcontainer/container_linux.go:1694]: The file contents, even if not synced to the disk, will be read back from the page cache the way it was written. The only thing that fsync helps with is when the host will lose power (or the kernel will crash) before writing the file to disk. I am not sure this is needed.

Comment (mrunalp) [reply libcontainer/container_linux.go]: Dropped the sync.

**Comment by mrunalp** [comment]
@cyphar Yeah, that's what I considered using first, but didn't want to depend for just a few lines of code and now with the sync being dropped, I am not sure if we want to pull that in, since that code does sync - https://github.com/moby/moby/blob/master/pkg/ioutils/fswriters.go#L68

**Pull request merged**